src/System.Net.Security/src/System/PinnableBufferCache.cs

```diff
@@ -0,0 +1,552 @@
+                        AgePendingBuffers();
+                        _trimmingExperimentInProgress = true;
+                        return true;
+                    }
+
+                    PinnableBufferCacheEventSource.Log.TrimFree(_cacheName, _buffersUnderManagement, freeCount, _notGen2.Count);
+                    _buffersUnderManagement -= _notGen2.Count;
+
+                    // Possibly revise the restocking down.  We don't want to grow aggressively if we are trimming.  
+                    var newRestockSize = _buffersUnderManagement / 4;
+                    if (newRestockSize < _restockSize)
+                        _restockSize = Math.Max(newRestockSize, DefaultNumberOfBuffers);
+
+                    _notGen2.Clear();
+                    _trimmingExperimentInProgress = false;
+                    return true;
+                }
+
+                // Set up an experiment where we use 25% less buffers in our free list.   We put them in 
+                // m_NotGen2, and if they are needed they will be put back in the free list again.  
+                var trimSize = freeCount / 4 + 1;
+
+                // We are OK with a 15% overhead, do nothing in that case.  
+                if (freeCount * 15 <= _buffersUnderManagement || _buffersUnderManagement - trimSize <= _minBufferCount)
+                {
+                    PinnableBufferCacheEventSource.Log.TrimFreeSizeOK(_cacheName, _buffersUnderManagement, freeCount);
+                    return true;
+                }
+
+                // Move buffers from the free list back to the non-aged list.  If we don't use them by next time, then we'll consider trimming them.
+                PinnableBufferCacheEventSource.Log.TrimExperiment(_cacheName, _buffersUnderManagement, freeCount, trimSize);
+                object buffer;
+                for (int i = 0; i < trimSize; i++)
+                {
+                    if (_freeList.TryPop(out buffer))
+                        _notGen2.Add(buffer);
+                }
+                _msecNoUseBeyondFreeListSinceThisTime = curMSec;
+                _trimmingExperimentInProgress = true;
+            }
+
+            // Indicate that we want to be called back on the next Gen 2 GC.  
+            return true;
+        }
+
+        private const int DefaultNumberOfBuffers = 16;
+        private string _cacheName;
+        private Func<object> _factory;
+
+        /// <summary>
+        /// Contains 'good' buffers to reuse.  They are guaranteed to be Gen 2 ENFORCED!
+        /// </summary>
+        private ConcurrentStack<object> _freeList = new ConcurrentStack<object>();
+        /// <summary>
+        /// Contains buffers that are not gen 2 and thus we do not wish to give out unless we have to.
+        /// To implement trimming we sometimes put aged buffers in here as a place to 'park' them
+        /// before true deletion.  
+        /// </summary>
+        private List<object> _notGen2;
```

Comment:
I don't understand why there's a preference for handing back things in gen2 over gen1 (or gen0).  Really there's a lot about this type that feels like it's trying to be clever in ways that it doesn't explain, which makes me wonder if it's engineered to last through the ages, or was just trying to match some performance goal from long ago and the universe has moved on making this now inefficient.

This feels like it should just be a ConcurrentQueue (or stack) and when empty it adds 10 (or whatever) and on a Gen2 collect it purges part of the cache if it's over some size; but no caring about what generation any particular object is in.